File: server/src/main/java/com/cyberbot/bomberman/server/SessionService.java
package com.cyberbot.bomberman.server;

import com.cyberbot.bomberman.core.models.net.InvalidPacketFormatException;
import com.cyberbot.bomberman.core.models.net.SerializationUtils;
import com.cyberbot.bomberman.core.models.net.packets.PlayerSnapshotPacket;
import com.cyberbot.bomberman.core.models.tiles.MissingLayersException;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.util.InvalidPropertiesFormatException;

public class SessionService implements GameSocket, Runnable {
    private final int port;
    private DatagramSocket socket;
    private volatile boolean sessionActive;
    private Session session;

    public SessionService(int port) throws MissingLayersException, InvalidPropertiesFormatException {
        this.port = port;
        this.sessionActive = true;
        this.session = new Session(this);
    }

    @Override
    public void run() {
        try {
            socket = new DatagramSocket(port);
            while (sessionActive) {
                byte[] buffer = new byte[1024];
                DatagramPacket packet = new DatagramPacket(buffer, buffer.length);
                socket.receive(packet);

                ClientConnection client = ClientConnection.fromDatagramPacket(packet);

                Object o = SerializationUtils.INSTANCE.deserialize(buffer, packet.getLength(), 0);
                if (!(o instanceof PlayerSnapshotPacket)) {
                    // TODO: Log warning
                    continue;
                }

                session.onSnapshot(client, (PlayerSnapshotPacket) o);
            }
        } catch (IOException | InvalidPacketFormatException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void send(DatagramPacket packet) throws IOException {
        socket.send(packet);
    }
}

File: server/src/main/java/com/cyberbot/bomberman/server/SessionService.kt
package com.cyberbot.bomberman.server

import com.cyberbot.bomberman.core.models.net.InvalidPacketFormatException
import com.cyberbot.bomberman.core.models.net.SerializationUtils.deserialize
import com.cyberbot.bomberman.core.models.net.packets.PlayerSnapshotPacket
import java.io.IOException
import java.net.DatagramPacket
import java.net.DatagramSocket

class SessionService(port: Int? = null, private val bufferSize: Int = 4096) : GameSocket, Runnable {
    val port: Int
    private val socket: DatagramSocket

    init {
        if (port == null) {
            socket = DatagramSocket()
            this.port = socket.port
        } else {
            socket = DatagramSocket(port)
            this.port = port
        }
    }

    @Volatile
    private var running = false
    private val session: Session = Session(this)

    override fun run() {
        try {
            running = true

            val buffer = ByteArray(bufferSize)
            val packet = DatagramPacket(buffer, buffer.size)
            while (running) {
                socket.receive(packet)
                val client = ClientConnection.fromDatagramPacket(packet)
                val snapshot = deserialize(buffer, packet.length, 0)
                        as? PlayerSnapshotPacket ?: continue
                session.onSnapshot(client, snapshot)
            }
        } catch (e: IOException) {
            e.printStackTrace()
        } catch (e: InvalidPacketFormatException) {
            e.printStackTrace()
        } finally {
            running = false
            socket.close()
        }
    }

    @Throws(IOException::class)
    override fun send(packet: DatagramPacket) {
        if (!running) {
            throw IllegalStateException("Attempting to send a packet when the service is not running")
        }

        socket.send(packet)
    }
}